grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RuleImpl.java

```diff
@@ -110,13 +109,6 @@ private Pattern parsePattern(String value){
         }
     }
 
-    /**
-     * Generate the internal hash in order to perform a faster lookups and ensure rules are unique
-     */
-    static String generateRuleIndex(Rule type, Pattern when, Pattern then){
-        return "RuleType_" + type.getLabel().getValue() + "_LHS:" + when.hashCode() + "_RHS:" + then.hashCode();
-    }
-
```

Comment:
So is there a performance hit for rule lookups since the index is removed? Was it a hash for the internal cache? I don't think I am personally using the get-rule-by-label option anyway.

Reply:
It was mostly used to ensure that you could not create two rules with the same pattern. But the protection was very artificial since simply using different variable names in the rules would let you bypass it.

Now we just removed that index and let duplicate rules be created. I.e. we allow the user error now.

Reply:
We could potentially add a validation rule for duplicates. Is there any gain in that? Reasoner recognises duplicates anyway.

Reply:
Hmmm we could. I was not aware reasoner could recognise dups. If that's the case we should use that mechanism at commit time. Can you show me where that code is and how I might repurpose it for commit time checks?

Reply:
I just had a look, so in theory you could compare bodies/heads by doing sth like:

```
ReasonerQuery body1 = Iterables.getOnlyElement(rule1.getWhen().admin().getDisjunctiveNormalForm().getPatterns()).toReasonerQuery(graph);
ReasonerQuery body2 = Iterables.getOnlyElement(rule2.getWhen().admin().getDisjunctiveNormalForm().getPatterns()).toReasonerQuery(graph);
```
and then simply doing

```
body1.equals(body2);
```

However the algorithm behind equals cannot recognise different variable binding patterns so two queries:

```
(chain)
($x, $y) isa some-relation;
($y, $z) isa some relation;
($z, $w) isa some-relation;
```
and

```
(loop)
($x, $y) isa some-relation;
($y, $z) isa some relation;
($z, $x) isa some-relation;
```
will be recognised as equal.

Reply:
Ah I already see an issue. `ReasonerQuery` is in `grakn-graql` commit time validation is done only with the presence of `grakn-kb` so we don't have access to those classes at commit time. 

We might need to invert the dependencies again

Reply:
`ReasonerQuery` is in `grakn-core`

Reply:
Oh just to be pedantic I will make that change in another PR